shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -619,6 +630,21 @@ - (UITextRange*)rangeEnclosingPosition:(UITextPosition*)position
 }
 
 - (UITextRange*)lineEnclosingPosition:(UITextPosition*)position {
+  // TODO(hellohuanlin): Remove iOS 17 check. The same logic should apply to older versions too.
+  if (@available(iOS 17.0, *)) {
+    // The end of document (with backward affinity) should not be part of any line, since end is
+    // exclusive. This is to fix 2 bugs:
```

Comment:
@chunhtai is "backward" equivalent to `TextAffinity.upstream`? If so I guess the "end is exclusive" statement isn't true I think?

Reply:
yes in **flutter** we use upstream and downstream to describe text affinity, and the upstream afinity is considered inclusive. not sure if they have different meaning or treated differently in iOS